python/pyarrow/tests/test_scalars.py

```diff
@@ -171,3 +171,37 @@ def test_dictionary(self):
                                            categorical.categories)
         for i, c in enumerate(values):
             assert v[i].as_py() == c
+
+    def test_int_hash(self):
+        # ARROW-640
+        arr = pa.array([1, 1, 2, 1])
+        arr2 = pa.array([1, 1, 2, 1])
+        assert arr[0].__hash__() == arr2[0].__hash__()
```

Comment:
Tests should call the built-in `hash()` function, not the `__hash__` method.
Also you need to check that hashing produces identical values to standard Python values, so for example `hash(arr[0]) == hash(1)`.

(if you were to write a custom hash implementation instead of delegating to `as_py()`, you would also need to test for many more int values, including "interesting" values like `-1`, `2**61`, etc.)

Reply:
Test have been updated. Noted on the custom hash. I think a custom hash faster than the built in is a larger task for me. I'd like to work on that, but thought it might be better to get a first pass hash implemented with the built in and then handle that separately.